engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41400to41500.java

```diff
@@ -235,6 +238,267 @@ private void updateSystemVmTemplates(final Connection conn) {
+                    pstmt.setLong(2, readOnlyUserSortOrder++);
+                    pstmt.executeUpdate();
+                }
+
+                if (rsRolePermissions != null && !rsRolePermissions.isClosed())  {
+                    rsRolePermissions.close();
+                }
+            }
+
+            if (rs != null && !rs.isClosed())  {
+                rs.close();
+            }
+            if (pstmt != null && !pstmt.isClosed())  {
+                pstmt.close();
+            }
+            LOG.debug("Successfully added role permissions for new read-only user role");
+        } catch (final SQLException e) {
+            LOG.error("Exception while adding role permissions for read-only user role: " + e.getMessage());
+            throw new CloudRuntimeException("Exception while adding role permissions for read-only user role: " + e.getMessage(), e);
+        }
+    }
+
+    private void addRolePermissionsForAdminSupport(final Connection conn) {
+        LOG.debug("Adding role permissions for new admin support role");
+        try {
+            PreparedStatement pstmt = conn.prepareStatement("SELECT id FROM `cloud`.`roles` WHERE name = 'Admin-Support (Default)' AND is_default = 1");
+            ResultSet rs = pstmt.executeQuery();
+            if (rs.next()) {
+                long adminSupportRoleId = rs.getLong(1);
+                int adminSupportSortOrder = 0;
+
+                pstmt = conn.prepareStatement("SELECT id FROM `cloud`.`roles` WHERE name = 'Read-Only Admin (Default)' AND is_default = 1");
+                ResultSet rsReadOnlyAdmin = pstmt.executeQuery();
+                if (rsReadOnlyAdmin.next()) {
+                    long readOnlyAdminRoleId = rsReadOnlyAdmin.getLong(1);
+                    pstmt = conn.prepareStatement("SELECT rule FROM `cloud`.`role_permissions` WHERE role_id = ? AND permission = 'ALLOW' ORDER BY sort_order");
+                    pstmt.setLong(1, readOnlyAdminRoleId);
+                    ResultSet rsRolePermissions = pstmt.executeQuery();
+
+                    while (rsRolePermissions.next()) {
+                        String rule = rsRolePermissions.getString(1);
+                        pstmt = conn.prepareStatement("INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`, `sort_order`) VALUES (UUID(), ?, ?, 'ALLOW', ?) ON DUPLICATE KEY UPDATE rule=rule");
+                        pstmt.setLong(1, adminSupportRoleId);
+                        pstmt.setString(2, rule);
+                        pstmt.setLong(3, adminSupportSortOrder++);
+                        pstmt.executeUpdate();
+                    }
+
+                    List<String> insertSqlForAdminSupportRolePermissions = new ArrayList<String>();
+                    insertSqlForAdminSupportRolePermissions.add("INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`, `sort_order`) VALUES (UUID(), ?, 'prepareHostForMaintenance', 'ALLOW', ?) ON DUPLICATE KEY UPDATE rule=rule");
```

Comment:
nit - Could be made less verbose by doing a loop on the API strings and add the string to the collection/list. Same for other parts of code.

Reply:
> nit - Could be made less verbose by doing a loop on the API strings and add the string to the collection/list. Same for other parts of code.

Updated with API rules